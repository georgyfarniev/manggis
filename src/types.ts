import { Connection, Document, Model, Error } from 'mongoose';

export type OnErrorHook = (ctx: IValidationError) => Promise<void>
export type OnModelHook = (model: Model<any>) => Promise<void>
export type OnDocumentHook = (doc: Document) => Promise<void>


// Validator execution context
// TODO: add defaults
export interface IValidationOptions {
  tempModelNamePrefix: string
  verifySchema: boolean
  verifyRefs: boolean

  // Fired once error received (per-field basis)
  onError: OnErrorHook

  // Fired once model processing started
  onModel?: OnModelHook

  // Fired once document processing started
  onDocument?: OnDocumentHook
}

export interface IValidationError {
  model: string
  document: string
  path: string
  message: string
}

export interface IValidationContext {
  options: IValidationOptions
  connection: Connection
  model?: Model<any>
  doc?: Document
  err?: Error.ValidationError
}
