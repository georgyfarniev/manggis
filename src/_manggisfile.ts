const ENTRYPOINT = '../../../aibolit/packages/common/dist/index.js'

// DB loading adapter stub!
export default async function getMongooseInstance() {
  const importedInstance = await import(ENTRYPOINT);

  const { connect, connection } = importedInstance.DB;

  await connect();

  return connection;
}