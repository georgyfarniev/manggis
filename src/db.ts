/**
 * Test database with mocks
 */

import mongoose, { Schema, Document, Types } from 'mongoose';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { ObjectId } from 'mongodb'

const mongod = new MongoMemoryServer();
// const DB_URL = 'mongodb://localhost:27017/manggis_test'

const NestedSchema: Schema = new Schema({
  field: { type: Number, required: true },
  bar: { type: Schema.Types.ObjectId, ref: 'Bar', required: true },
});

const FooSchema: Schema = new Schema({
  name: { type: String, required: true },
  bar: { type: Schema.Types.ObjectId, ref: 'Bar', required: true },
  barOptional: { type: Schema.Types.ObjectId, ref: 'Bar', required: false },
  subdoc: NestedSchema,
  subarray: [NestedSchema],
  refarray: [{ type: Schema.Types.ObjectId, ref: 'Bar', required: true }]
});

const Foo = mongoose.model('Foo', FooSchema);

const BarSchema: Schema = new Schema({
  title: { type: String, required: true }
});

const Bar = mongoose.model('Bar', BarSchema);

export async function createTestDatabase() {
  const DB_URL = await mongod.getUri();

  await mongoose.connect(DB_URL, {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  const [bar1, bar2] = await Bar.create([
    { title: 'bar1' },
    { title: 'bar2' }
  ], [])

  const nonexistingId = new ObjectId();

  await Foo.collection.insertMany([
    // correct
    {
      name: 'foo1',
      bar: bar1._id,
      subdoc: { field: 111, bar: bar2._id },
      subarray: [{ field: 222, bar: bar1._id }],
      refarray:[nonexistingId]
    },

    // corrrect with optional field set as well
    {
      name: 'foo2',
      bar: bar1._id,
      barOptional: bar2._id 
    },

    // incorrect (required, but missing), will trigger validation error
    {
      name: 'foo3',
      bar: null
    },

    // incorrect, broken reference (nonexisting target)
    {
      name: 'foo3',
      bar: nonexistingId
    }
  ]);

  console.log('db connected', DB_URL);
  return mongoose.connection;
}