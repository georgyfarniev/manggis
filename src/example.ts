
import { Connection } from 'mongoose';
import { validate } from '.'

async function main() {
  /* !!! place your mongoose connection instance here !!! */
  const connection: Connection = {} as Connection

  await validate(connection, {
    onError: ({ model, path, message }) => {
      console.log(`    error [${model}.${path}]: ${message}`)
    },
    onModel(model) {
      console.log('validating model:', model.modelName);
    },
    onDocument(doc) {
      console.log('  validating document:', doc._id);
    }
  });

  console.log('all models has been verified');
  process.exit();
}

if (require.main === module) {
  main()
}
