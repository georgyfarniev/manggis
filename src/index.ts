import assert from 'assert';
import mongoose, { Connection, Schema } from 'mongoose';
import type { IValidationContext, IValidationOptions } from './types';

/**
 * TODO list:
 * 1. Optionally allow to specify query filters (by exposing lower level api)
 * 2. Unit tests
 * 3. Advertise it in blogs or social networks
 * 4. Allow hooks to cancel execution of certain tests
 */


const DEFAULT_OPTIONS: IValidationOptions = {
  tempModelNamePrefix: '__manggis_model__',
  verifyRefs: true,
  verifySchema: true
}

export async function validate(connection: Connection, opts: IValidationOptions) {
  const options: IValidationOptions = { ...opts, ...DEFAULT_OPTIONS };

  assert(options.verifyRefs || options.verifySchema, 'at least one validation option is a must');

  const ctx: IValidationContext = {
    connection,
    options,
  }

  const models = Object.values(connection.models);

  for (const model of models) {
    // Only for refs need to create custom model
    const validationModel = options.verifyRefs ? createEnhancedModel({ ...ctx, model }) : model;

    if (options.onModel) await options.onModel(model);

    const cursor = validationModel.find();
    for await (const doc of cursor) {
      if (options.onDocument) await options.onDocument(doc);

      if (options.verifySchema) {
        const docCtx = { ...ctx, model: validationModel, doc }
        await verifySchemaForDocument(docCtx);
      }
    }

    if (options.verifyRefs) {
      mongoose.deleteModel(validationModel.modelName)
    }
  }
}

function createEnhancedModel(ctx: IValidationContext) {
  const schema = ctx.model!.schema.clone();

  const { tempModelNamePrefix, verifyRefs } = ctx.options

  if (verifyRefs) {
    installRefValidator(ctx.connection, schema);
  }

  const name = `${tempModelNamePrefix}${ctx.model!.modelName}`
  return mongoose.model(name, schema, ctx.model!.collection.collectionName);
}

async function verifySchemaForDocument(ctx: IValidationContext) {
  try {
    await ctx.doc!.validate();
  } catch (err) {
    if (err instanceof mongoose.Error.ValidationError) {
      await reportValidationErrors({ ...ctx, err });
    } else {
      assert(false, 'Unknown error!');
    }
  }
}

async function reportValidationErrors(ctx: IValidationContext) {
  const { options, doc, err } = ctx;
  const { onError, tempModelNamePrefix } = options;

  if (!onError) return;

  const model = ctx.model!.modelName.replace(tempModelNamePrefix!, '');

  for (const error of Object.values(err!.errors)) {
    await onError({
      document: doc!._id.toString(),
      model,
      path: error.path,
      message: error.message
    })
  }
}

function installRefValidator(connection: Connection, schema: Schema) {
  // paths are not fully typed yet!
  for (const type of Object.values<any>(schema.paths)) {
    if (type.schema) {
      installRefValidator(connection, type.schema);
    } else if (type.options.ref) {
      type.validate({
          validator: async (value: any) => {
            const { ref } = type.options;
            const refModel = connection.models[ref];

            assert(refModel, `Referenced model not found: ${ref}`);

            const exists = await refModel.exists({ _id: value });
            return exists;
          },
          message: (props: any) => `Referenced document not found: ${props.value}`
      })
    }
  }
}
