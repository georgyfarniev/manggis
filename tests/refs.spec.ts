import 'jest'
import { Document, Model, Connection } from 'mongoose'
import { createEnhancedModel, verifySchemaForDocument } from '../src/'
import { IValidationContext, IValidationError, IValidationOptions } from '../src/types';
import { createTestDatabase, destroyTestDatabase } from './_db'

async function wrapModel(connection: Connection, model: Model<any>): Promise<Model<any>> {
  const options: IValidationOptions = {
    verifyRefs: true,
    verifySchema: true,
  };

  const wrappedModel = createEnhancedModel({
    connection,
    model,
    options
  });

  return wrappedModel;
}

async function verifyDoc(connection: Connection, model: Model<any>, obj: any) {
  const errors: IValidationError[] = [];

  const options: IValidationOptions = {
    verifyRefs: true,
    verifySchema: true,
    onError(err) { errors.push(err); }
  };

  const doc = new model(obj);

  await verifySchemaForDocument({
    connection,
    model,
    options,
    doc
  });

  return errors;
}

describe('Refs validation', () => {
  let connection: Connection

  let FooWrapper: Model<any>
  let BarWrapper: Model<any>

  beforeAll(async () => {
    connection = await createTestDatabase();
    FooWrapper = await wrapModel(connection, connection.models.Foo);
    BarWrapper = await wrapModel(connection, connection.models.Bar);
  })

  afterAll(async () => {
    await destroyTestDatabase();
  });

  test('db sanity test', async () => {
    expect(FooWrapper).toBeDefined();
    expect(BarWrapper).toBeDefined();

    await expect(FooWrapper.count()).resolves.toBeGreaterThan(0);
  })

  test('advanced test', async () => {

    const errors = await verifyDoc(connection, FooWrapper, {
      name: 'foo3',
      bar: null
    });

    expect(errors).toMatchSnapshot();
  })
})

