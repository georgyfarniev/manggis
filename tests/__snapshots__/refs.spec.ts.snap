// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refs validation advanced test 1`] = `
Array [
  Object {
    "document": "604c44c8bb7e660d2c36e4c8",
    "message": "Path \`bar\` is required.",
    "model": "Foo",
    "path": "bar",
  },
]
`;
